### ISSUE TITLE
BufferedProxySelector - Caching with individual URLs might not be ideal

### ISSUE THREAD

**Issue by techpavan** [MarkusBernhardt/proxy-vole#1]
Hi,

The below comment mentions that caching per host may produce wrong results. I am not sure about a scenario where we can expect wrong results with that. Please clarify if any such possible scenarios can exist, or would it be preferable to modify it to cache based on the host:port combination?

https://github.com/MarkusBernhardt/proxy-vole/blob/master/src/main/java/com/github/markusbernhardt/proxy/selector/misc/BufferedProxySelector.java#L89

**Comment by MarkusBernhardt** [comment]
Hi,

the problem here is PAC. A proxy auto-config (PAC) file defines how web browsers and other user agents can automatically choose the appropriate proxy server (access method) for fetching a given URL. A PAC file contains a JavaScript function “FindProxyForURL(url, host)” and can deliver separate proxies for different sub-URLs. Very esoteric, but possible.

Is the current implementation a problem for you?

Best regards,
Markus

**Comment by MarkusBernhardt** [comment]
Hi techpavan,
any updates?

**Comment by techpavan** [comment]
Hi Markus,

I feel it just reduces the efficiency of the invoking application by introducing more number of IO reads. Say, a request is made for example.com/abc - this involves an IO operation to read the PAC file, it might not be justified to make one more IO when a request is received for example.com/def.

Though, from spec, we do not see a clarity on whether two urls from same domain can resolve through different proxies - it is just not realistic to access two urls from same server through different proxies.

With these, BufferedProxySelector is implemented with a strict view of ensuring accuracy at all times. But in most of the scenarios, such strict evaluation at the cost of IO might not be necessary. Probably you might consider providing a flag to the consumer to avoid such strict check and just use host, port based caching and thus speed up future lookups.

**Note:** I solved it in my application by stripping off the file part of the url while making a request to the selector. Thus, even if request is for /abc or /def, request to selector is always just "http://example.com"

**Note2:** Microsoft IE also does caching and they do it based on host. They provide a switch for the user to turn off if such caching causes an issue in the user's environment - https://support.microsoft.com/en-in/kb/271361

Thanks,
Pavan

**Comment by MarkusBernhardt** [comment]
Hi Pavan,

I added now a scope to the cache. The default will be to cache based on hosts as the Internet Explorer. Could you please review the change.

Thanks,
Markus

**Comment by MarkusBernhardt** [comment]
Fixed with release 1.0.2

**Comment by techpavan** [comment]
Hi Markus,

Thanks for getting it fixed so quick. From a high level diff, it seems like a public method and constructor are modified. This might cause a problem for someone who might just want to upgrade as the backward compatibility is broken. Its preferable to introduce new constructor / method rather than updating the signatures.

Thanks,
Pavan

**Comment by MarkusBernhardt** [comment]
Yes, you are right. I changed the constructor, because I did not want to have the default cache scope defined at two places. So far the pac cache defaults are in ProxySearch. Didn't like to change that.

I was also thinking about introducing a new method, something like "setPacCacheScope" and leave "setPacCacheSettings" untouched or simply overload "setPacCacheSettings", but I belive the change is so trivial it will not prevent users from adapting to the new Version.
